### PR TITLE
Make ASAN builds work again

### PR DESCRIPTION
**Pull request by dioni21**
### Motivation and Context

Commit af65916 added -Wl,-z,defs for the shared libraries. This apparently does not work in some cases with --enable-asan.

Does not work in fc31 and fc32, but works in gentoo.

Closes: https://github.com/openzfs/zfs/issues/10557

### Description

Conditionally, so only disable changes for ASAN builds.

### How Has This Been Tested?

Build ASAN now works again.

### Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [X] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by dioni21** [comment]
@nivedita76 could you please review this?

**Comment by nivedita76** [comment]
> @nivedita76 could you please review this?

I commented on the issue. I don't actually get errors gcc-9.3.0 and gcc-10.1.0. They link in libasan and there are no undefined symbols in the shared libs.

**Comment by nivedita76** [comment]
> > @nivedita76 could you please review this?
> 
> I commented on the issue. I don't actually get errors gcc-9.3.0 and gcc-10.1.0. They link in libasan and there are no undefined symbols in the shared libs.

The change itself looks ok though I'd like to understand why it's working for me and not for you. I'd prefer the formatting in https://github.com/nivedita76/zfs/commit/3da0bae04b7a4b9595fd61fc9ce386ede004cfd2 but it's not a big deal.

**Comment by dioni21** [comment]
> > @nivedita76 could you please review this?
> 
> I commented on the issue. I don't actually get errors gcc-9.3.0 and gcc-10.1.0. They link in libasan and there are no undefined symbols in the shared libs.

Maybe the problem is in Fedora 31?

```
Using built-in specs.
COLLECT_GCC=/bin/gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC) 
```

**Comment by dioni21** [comment]
@nivedita76:

> The change itself looks ok though I'd like to understand why it's working for me and not for you. I'd prefer the formatting in [nivedita76@3da0bae](https://github.com/nivedita76/zfs/commit/3da0bae04b7a4b9595fd61fc9ce386ede004cfd2) but it's not a big deal.

Look better, indeed.  Brought it to this PR, to keep it simpler

**Comment by nivedita76** [comment]
> @nivedita76:
> 
> > I don't actually get errors gcc-9.3.0 and gcc-10.1.0. They link in libasan and there are no undefined symbols in the shared libs.
> 
> Maybe the problem is in Fedora? I tested with fc31 and fc32.
> 
> ```
> Using built-in specs.
> COLLECT_GCC=/bin/gcc
> COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/9/lto-wrapper
> OFFLOAD_TARGET_NAMES=nvptx-none
> OFFLOAD_TARGET_DEFAULT=1
> Target: x86_64-redhat-linux
> Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
> Thread model: posix
> gcc version 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC) 
> ```

what does gcc -dumpspecs show? I think the asan-related bit is in `*link_command:`, mine looks like this:
`... %{!nostdlib:%{!r:%{!nodefaultlibs:%{%:sanitize(address):%{!shared:libasan_preinit%O%s} %{static-libasan:%{!shared:-Bstatic --whole-archive -lasan --no-whole-archive -Bdynamic}}%{!static-libasan:-lasan}} ...`

File: lib/libnvpair/Makefile.am [hunk 1]
```diff
@@ -31,10 +31,14 @@ libnvpair_la_LIBADD = \
...
+if !ASAN_ENABLED
+libnvpair_la_LDFLAGS += -Wl,-z,defs
+endif
 
 if BUILD_FREEBSD
-libnvpair_la_LDFLAGS += -version-info 3:0:0
+libnvpair_la_LDFLAGS = -version-info 3:0:0
 else
-libnvpair_la_LDFLAGS += -version-info 1:1:0
+libnvpair_la_LDFLAGS = -version-info 1:1:0
```
Comment (behlendorf) [inline lib/libnvpair/Makefile.am:43]: These need to still use `+=`.

Comment (nivedita76) [reply lib/libnvpair/Makefile.am:43]: Yeah looks like that's left-over from the original patch.

Comment (dioni21) [reply lib/libnvpair/Makefile.am]: Done

**Comment by dioni21** [comment]
@nivedita76:
> what does gcc -dumpspecs show? I think the asan-related bit is in `*link_command:`, mine looks like this:

Never seen this before, hope this is the correct info:
```
...
%{!nostdlib:%{!r:%{!nodefaultlibs:%{%:sanitize(address):%{!shared:libasan_preinit%O%s} %{static-libasan:%{!shared:-Bstatic --whole-archive -lasan --no-whole-archive -Bdynamic}}%{!static-libasan:-lasan}}
...
%{!nostdlib:%{!r:%{!nodefaultlibs:%{%:sanitize(address): %{static-libasan|static:%:include(libsanitizer.spec)%(link_libasan)}    %{static:%ecannot specify -static with -fsanitize=address}}    %{%:sanitize(thread): %{static-libtsan|static:%:include(libsanitizer.spec)%(link_libtsan)}    %{static:%ecannot specify -static with -fsanitize=thread}}    %{%:sanitize(undefined):%{static-libubsan:-Bstatic} -lubsan %{static-libubsan:-Bdynamic} %{static-libubsan|static:%:include(libsanitizer.spec)%(link_libubsan)}}    %{%:sanitize(leak): %{static-liblsan|static:%:include(libsanitizer.spec)%(link_liblsan)}}}}}
...
```

**Comment by nivedita76** [comment]
> @nivedita76:
> 
> > what does gcc -dumpspecs show? I think the asan-related bit is in `*link_command:`, mine looks like this:
> 
> Never seen this before, hope this is the correct info:
> 
> ```
> ...
> %{!nostdlib:%{!r:%{!nodefaultlibs:%{%:sanitize(address):%{!shared:libasan_preinit%O%s} %{static-libasan:%{!shared:-Bstatic --whole-archive -lasan --no-whole-archive -Bdynamic}}%{!static-libasan:-lasan}}
> ...
> %{!nostdlib:%{!r:%{!nodefaultlibs:%{%:sanitize(address): %{static-libasan|static:%:include(libsanitizer.spec)%(link_libasan)}    %{static:%ecannot specify -static with -fsanitize=address}}    %{%:sanitize(thread): %{static-libtsan|static:%:include(libsanitizer.spec)%(link_libtsan)}    %{static:%ecannot specify -static with -fsanitize=thread}}    %{%:sanitize(undefined):%{static-libubsan:-Bstatic} -lubsan %{static-libubsan:-Bdynamic} %{static-libubsan|static:%:include(libsanitizer.spec)%(link_libubsan)}}    %{%:sanitize(leak): %{static-liblsan|static:%:include(libsanitizer.spec)%(link_liblsan)}}}}}
> ...
> ```

That looks the same as mine. Dunno what else could be wrong.

**Comment by behlendorf** [comment]
@nivedita76 unless you'd like to dig in this further, I'd suggest we merge this fix as it stands.

**Comment by nivedita76** [comment]
> @nivedita76 unless you'd like to dig in this further, I'd suggest we merge this fix as it stands.

Yes we should definitely merge this now. We can try to figure it out later anyway.

**Pull request merged**